src/localQueue.ts

```diff
@@ -0,0 +1,415 @@
+
+  /**
+   * For promises that happen in the background, but that we want to ensure are
+   * handled before we release the queue (so that the database pool isn't
+   * released too early).
+   */
+  private background(promise: Promise<void>) {
+    if (this.mode === "RELEASED" && this.backgroundCount === 0) {
+      throw new Error(
+        `Cannot background something when the queue is already released`,
+      );
+    }
+    this.backgroundCount++;
+    promise.then(this.decreaseBackgroundCount, this.decreaseBackgroundCount);
+  }
+
+  private setModePolling() {
+    assert.ok(
+      !this.fetchTimer,
+      "Cannot enter polling mode when a fetch is scheduled",
+    );
+    assert.ok(
+      !this.fetchInProgress,
+      "Cannot enter polling mode when fetch is in progress",
+    );
+    assert.equal(
+      this.jobQueue.length,
+      0,
+      "Cannot enter polling mode when job queue isn't empty",
+    );
+
+    if (this.ttlExpiredTimer) {
+      clearTimeout(this.ttlExpiredTimer);
+      this.ttlExpiredTimer = null;
+    }
+
+    this.mode = POLLING;
+
+    this.fetch();
+  }
+
+  private setModeWaiting() {
+    // Can only enter WAITING mode from POLLING mode.
+    assert.equal(this.mode, POLLING);
+    assert.ok(
+      !this.fetchTimer,
+      "Cannot enter waiting mode when a fetch is scheduled",
+    );
+    assert.ok(
+      !this.fetchInProgress,
+      "Cannot enter waiting mode when fetch is in progress",
+    );
+    assert.notEqual(
+      this.jobQueue.length,
+      0,
+      "Cannot enter waiting mode when job queue is empty",
+    );
+
+    if (this.ttlExpiredTimer) {
+      clearTimeout(this.ttlExpiredTimer);
+    }
+
+    this.mode = WAITING;
+
+    this.ttlExpiredTimer = setTimeout(() => {
+      this.setModeTtlExpired();
+    }, this.ttl);
+  }
+
+  private setModeTtlExpired() {
+    // Can only enter TTL_EXPIRED mode from WAITING mode.
+    assert.equal(this.mode, WAITING);
+    assert.ok(
+      !this.fetchTimer,
+      "Cannot enter TTL expired mode when a fetch is scheduled",
+    );
+    assert.ok(
+      !this.fetchInProgress,
+      "Cannot enter TTL expired mode when fetch is in progress",
+    );
+    assert.notEqual(
+      this.jobQueue.length,
+      0,
+      "Cannot enter TTL expired mode when job queue is empty",
+    );
+
+    if (this.ttlExpiredTimer) {
+      clearTimeout(this.ttlExpiredTimer);
+      this.ttlExpiredTimer = null;
+    }
+
+    this.mode = TTL_EXPIRED;
+
+    // Return jobs to the pool
+    this.returnJobs();
+  }
+
+  private returnJobs() {
+    const jobsToReturn = this.jobQueue.splice(0, this.jobQueue.length);
+    this.background(
+      returnJob(
+        this.compiledSharedOptions,
+        this.withPgClient,
+        this.workerPool.id,
+        jobsToReturn,
+      ).then((e) => {
+        // TODO: handle this better!
+        this.compiledSharedOptions.logger.error(
+          `Failed to return jobs from local queue to database queue`,
+          { error: e },
+        );
```

Comment:
@benjie  I am seeing this error in a local test - shouldn't this be `catch` instead of `then`?

Reply:
Uhhh... yeah I think so!